037-pluggable-pod-security-profiles.md

```diff
@@ -0,0 +1,196 @@
+These methods will have the default implementation which will always just return the user-specified context.
+For example:
+
+```java
+    default PodSecurityContext kafkaPodSecurityContext(PodSecurityContext userSecurityContext, Storage storage) {
+        return userSecurityContext;
+    }
+
+    default SecurityContext kafkaContainerSecurityContext(SecurityContext userSecurityContext, Storage storage) {
+        return userSecurityContext;
+    }
+
+    default SecurityContext kafkaInitContainerSecurityContext(SecurityContext userSecurityContext) {
+        return userSecurityContext;
+    }
+```
+
+The interface will also have a `configure` method for configuring the provider.
+The `configure` method will get as a parameter an object describing the platform (Kubernetes version, features / APIs) which can be taken into account when creating the security context.
+This method will not havy any default implementtion.
+
+#### Implementations
+
+We will also provide two implementations:
+* `BaselinePodSecurityProvider`
+* `RestrictedPodSecurityProvider`
+
+These implementations will be also part of the `api` module so that users can just extend them instead of implementing the provide from scratch.
+
+The `BaselinePodSecurityProvider` will implement the current behaviour:
+* Return the user-configured security context when specified
+* When the user-configure context is `null` and we are on OpenShift, a `null` will be returned (OpenShift injects its own Security Context in that case) for all Pods and containers
+* When the user-configure context is `null` and we are not on OpenShift:
+    * Pods with storage will get a Security Context specifying the file system group as `0`
+    * All other pods will get `null` Security Context
+    * All containers will get `null` Security Context
+
+The `RestrictedPodSecurityProvider` will implement the following behaviour:
+* Return the user-configured security context when specified
+* When the user-configure context is `null`:
+    * Pods with storage will get a Security Context specifying the file system group as `0`
+    * All other pods will get `null` Security Context
+    * For the Kaniko container, an exception will be thrown since Kaniko cannot run under the Restricted profile
+    * All other containers will get a Security Context matching the _restricted_ Kubernetes profile:
+        * Disable privilage escalation
+        * Run as non-root
+        * Specify the Seccomp profile as `RuntimeDefault`
+        * Drop all capabilities
+
+#### Loading the plugins
+
+In the Cluster Operator, the plugins will be loaded using the Java Service Provider Interface which will be used to find the class configured by the user.
+A `PodSecurityProviderFactory` class will be part of the `cluster-oprator` module.
+It will have a static initialize method which will load the class based on the configuration, call its `configure(...)` method and store it in a static field for further use.
```

Comment:
Can you should the `configure` method in the example above?

Reply:
Done.